### DISCUSSION TITLE
RDS1 'Long PS' support.

### DISCUSSION THREAD

**Discussion by AaronJowitt** [PE5PVB/TEF6686_ESP32#566] [Ideas]
RDS1 supports Long PS in Group 15A (IEC RDS1 standard 62106-2).  There are a number of  transmitters in this area broadcasting LPS (in addition to PS).

Most radios supporting LPS typically replace the PS display with LPS.  Would be a nice addition to the TEF for the advanced RDS screen, and a perhaps a PS/ scrolling LPS menu toggle option for the home screen.

**Comment by PE5PVB** [comment]
Can you provide me some MPX recordings for debugging?

**Reply by AaronJowitt** [reply]
Sure.  Here is a 4 minute file via WeTransfer.  Station broadcasts a static PS &  static LPS. It is also transmitting the station logo on a RDS2 sub carrier.

MPX 50us

https://we.tl/t-6tN2NBi6ca

**Reply by PE5PVB** [reply]
Didn't debug it, but please test this version: https://we.tl/t-3wukmhWOsX

**Reply by AaronJowitt** [reply]
Thanks Sjef. Tried that and now RC4.

All good except there is no space/s between the last character of the LPS and the first character when loop scrolling.

**Reply by PE5PVB** [reply]
Fixed

**Reply by AaronJowitt** [reply]
Fix confirmed.  Thank you!

**Comment by PE5PVB** [comment]
OK. I have a day off today so I can play with it. The TEF can't decode RDS2 but I think I was able to parse the long PS.

Is the text 'Breeze 100.6 Gold Coast' correct?

Maybe you can provide more Long PS recordings to do double checks?

**Reply by AaronJowitt** [reply]
Yes that's it -  the Long PS is 'Breeze 100.6 Gold Coast'.

 I can record a different station with a static LPS if you would like?

**Reply by PE5PVB** [reply]
Yes please, because I have no sources.

**Reply by AaronJowitt** [reply]
Sure. This station has a static LPS of "REBEL 99.4 Gold Coast".  And broadcasts a station logo via RDS2.  If of interest, latest RDS SPY with the recent Pira RDS SPY Plugin will decode/display the logo (takes over a minute).

MPX 50us

https://we.tl/t-tiGCxp5yBl

**Comment by PE5PVB** [comment]
Done, see https://github.com/PE5PVB/TEF6686_ESP32/commit/b228eb63c4da8ba1ac3ebf82a89afd3067543db9

Do you have access to the Discord server? I'll post a new test version in an hour.

**Comment by PE5PVB** [comment]
I just posted it on Discord v2.10 RC2

**Reply by AaronJowitt** [reply]
Thanks!

**Reply by AaronJowitt** [reply]
The MPX files I provided above have NULs in Group 15A padding out the LPS text to 32 characters.  

The LPS encoding/transmission has now been changed to have no text padding and is terminated by 1-4 CR's. If you require a new MPX file reflecting this newer method, please let me know.

TEF still displays the LPS ok, but noticed LPS only displays if FAST PS is set to ON. If set OFF only standard PS is shown.  Is that intentional? Just curious.

**Reply by PE5PVB** [reply]
Can you provide a new file for testing?

**Reply by PE5PVB** [reply]
BTW. What do you mean with NULs? Is this 0x00? 0x00 will be ignored by the RDS parser.

So in previous version all 32 characters were used and not used chars at the end of the text were all 0x00? Now these blocks are not transmitted and text is ended with 0x0d. Correct?

If so, I already implemented that, but I found a bug that when not all blocks were transmitted and 0x0d was present it never goes to slow PS mode. Just fixed that.

**Reply by AaronJowitt** [reply]
Yes correct. There may be 1 to 4 CR's (0x0d) at the end. Here is a new MPX file just recorded that reflects that. 

Link expires in 3 days: 
https://we.tl/t-o3BHbd1PNx

**Comment by andimik** [comment]
Thanks for the file.

`redsea` does not support it (yet), I'll tell Oona to implement it.

```
{"pi":"0x49B1","group":"15A","prog_type":"Easy listening","tp":false,"unknown_oda":{"raw_data":"00 4272 6565"}}
{"pi":"0x49B1","group":"15A","prog_type":"Easy listening","tp":false,"unknown_oda":{"raw_data":"01 7A65 2031"}}
{"pi":"0x49B1","group":"15A","prog_type":"Easy listening","tp":false,"unknown_oda":{"raw_data":"02 3030 2E36"}}
{"pi":"0x49B1","group":"15A","prog_type":"Easy listening","tp":false,"unknown_oda":{"raw_data":"03 2047 6F6C"}}
{"pi":"0x49B1","group":"15A","prog_type":"Easy listening","tp":false,"unknown_oda":{"raw_data":"04 6420 436F"}}
{"pi":"0x49B1","group":"15A","prog_type":"Easy listening","tp":false,"unknown_oda":{"raw_data":"05 6173 7400"}}
{"pi":"0x49B1","group":"15A","prog_type":"Easy listening","tp":false,"unknown_oda":{"raw_data":"06 0000 0000"}}
{"pi":"0x49B1","group":"15A","prog_type":"Easy listening","tp":false,"unknown_oda":{"raw_data":"07 0000 0000"}}
```

```
$ redsea -f Downloads/Breeze\ 100.6\ MPX\ 17May2024.wav -x | grep F18 | head
49B1 F187 0000 0000
49B1 F180 4272 6565
49B1 F181 7A65 2031
49B1 F182 3030 2E36
49B1 F183 2047 6F6C
49B1 F184 6420 436F
49B1 F185 6173 7400
49B1 F186 0000 0000
49B1 F187 0000 0000
49B1 F180 4272 6565
```

**Comment by andimik** [comment]
Are you the owner of that station? I have never seen such a clean MPX recording ;-)

![grafik](https://github.com/PE5PVB/TEF6686_ESP32/assets/24510556/da484994-ef8a-490b-8b8b-e442c41fa557)

**Reply by AaronJowitt** [reply]
yes, thanks. That is the feed to the tx.

**Comment by AaronJowitt** [comment]
Works well thanks.  A listener posted a YouTube video.

https://www.youtube.com/watch?v=kHvE3c83Luc&ab_channel=TheAndyShow
